GameServerLib/Maps/Surrender.cs

```diff
@@ -0,0 +1,113 @@
+using System.Linq;
+
+namespace LeagueSandbox.GameServer.Maps
+{
+    // TODO: Make the surrender UI button become clickable upon hitting SurrenderMinimumTime
+    public class Surrender : IUpdate
+    {
+        private Dictionary<IChampion, bool> _votes = new Dictionary<IChampion, bool>();
+        private Game _game;
+
+        public float SurrenderMinimumTime { get; set; }
+        public float SurrenderRestTime { get; set; }
+        public float SurrenderLength { get; set; }
+        public float LastSurrenderTime { get; set; }
+        public bool IsSurrenderActive { get; set; }
+        public TeamId Team { get; set; }
+
+        // TODO: The first two parameters are in milliseconds, the third is seconds. QoL fix this?
+        public Surrender(Game g, TeamId team, float minTime, float restTime, float length)
+        {
+            _game = g;
+            Team = team;
+            SurrenderMinimumTime = minTime;
+            SurrenderRestTime = restTime;
+            SurrenderLength = length;
+        }
+
+        public Pair<int, int> GetVoteCounts()
+        {
+            int yes = _votes.Count(kv => kv.Value == true);
+            int no = _votes.Count - yes;
+            return new Pair<int, int>(yes, no);
+        }
+
+        public void HandleSurrender(int userId, IChampion who, bool vote)
+        {
+            if (_game.GameTime < SurrenderMinimumTime)
+            {
+                _game.PacketNotifier.NotifySurrenderStatus(userId, who.Team, SurrenderReason.SURRENDER_TOO_EARLY, 0, 0);
+                return;
+            }
+            
+            bool open = !IsSurrenderActive;
+            if (!IsSurrenderActive)
+            {
+                if (_game.GameTime < LastSurrenderTime + SurrenderRestTime)
+                {
+                    _game.PacketNotifier.NotifySurrenderStatus(userId, who.Team, SurrenderReason.SURRENDER_TOO_QUICKLY, 0, 0);
+                    return;
+                }
+                IsSurrenderActive = true;
+                LastSurrenderTime = _game.GameTime;
+                _votes.Clear();
+            }
+
+            if (_votes.ContainsKey(who))
+            {
+                _game.PacketNotifier.NotifySurrenderStatus(userId, who.Team, SurrenderReason.SURRENDER_ALREADY_VOTED, 0, 0);
+                return;
+            }
+            _votes[who] = vote;
+            Pair<int, int> voteCounts = GetVoteCounts();
+            int total = _game.PlayerManager.GetPlayers().Count;
+
+            Console.WriteLine($"Champion {who.Model} voted {vote}. Currently {voteCounts.Item1} yes votes, {voteCounts.Item2} no votes, with {total} total players");
+
+            _game.PacketNotifier.NotifySurrender(who, open, vote, (byte)voteCounts.Item1, (byte)voteCounts.Item2, (byte)total, SurrenderLength);
+
+            if (voteCounts.Item1 >= total - 1)
+            {
+                IsSurrenderActive = false;
+                foreach (Pair<uint, ClientInfo> p in _game.PlayerManager.GetPlayers())
```

Comment:
Don't use Pair (see #780 )

Here this is tricky cause the whole system of `PlayerManager` use `Pair`. I recommend you will change to:
`var p in _game.PlayerManager.GetPlayers().Item1`
So you will only get the player id.

Reply:
I'll most likely make a new PR to refactor using `Tuple` instead of `Pair`, however it's used a lot in all projects. I'll keep this `Pair` but use your suggested change until I get around to refactoring all in a new PR.